File: Longest String Chain/Longest_String_Chain.cpp
class Solution {
  public:
    bool check(string s,string x) {
        int n = s.size();
        int m = x.size();
        int i = 0,j = 0;
        int cnt = 0;
        while(i<n && j<m) {
            if(s[i] == x[j]) {
                i++;
                j++;
            }
            else {
                i++;
                cnt++;
            }
        }
        if(j<m) return false;
        while(i<n) {
            i++;
            cnt++;
        }
        return cnt == 1;
    }
    int longestStringChain(vector<string>& words) {
        map<int,vector<pair<string,int>>> mpp;
        for(auto i: words) {
            mpp[i.size()].push_back({i,1});
        }
        int ans = 0;
        for(auto &i: mpp) {
            for(auto &x: i.second) {
                for(auto j: mpp[i.first-1]) {
                    if(check(x.first,j.first)) {
                        x.second = max(j.second+1,x.second);
                    }
                }
                ans = max(ans,x.second);
            }
        }
        return ans;
    }
};

File: Longest String Chain/Longest_String_Chain.java
class Pair<K, V> {
    private final K key;
    private final V value;
    
    public Pair(K key, V value) {
        this.key = key;
        this.value = value;
    }
    
    public K getKey() {
        return key;
    }
    
    public V getValue() {
        return value;
    }
}

class Solution {
    private boolean check(String s, String x) {
        int n = s.length();
        int m = x.length();
        int i = 0, j = 0, cnt = 0;
        while (i < n && j < m) {
            if (s.charAt(i) == x.charAt(j)) {
                i++;
                j++;
            } else {
                i++;
                cnt++;
            }
        }
        if (j < m) return false;
        while (i < n) {
            i++;
            cnt++;
        }
        return cnt == 1;
    }
    
    public int longestStringChain(String words[]) {
        Map<Integer, List<String>> mpp = new HashMap<>();
        Map<String, Integer> dp = new HashMap<>();
        for (String word : words) {
            mpp.putIfAbsent(word.length(), new ArrayList<>());
            mpp.get(word.length()).add(word);
            dp.put(word, 1);
        }
        int ans = 0;
        for (Map.Entry<Integer, List<String>> entry : mpp.entrySet()) {
            for (String x : entry.getValue()) {
                for (String j : mpp.getOrDefault(entry.getKey() - 1, new ArrayList<>())) {
                    if (check(x, j)) {
                        dp.put(x, Math.max(dp.get(j) + 1, dp.get(x)));
                    }
                }
                ans = Math.max(ans, dp.get(x));
            }
        }
        return ans;
    }
}